### ISSUE TITLE
Copilot does not work on a continous basis

### ISSUE THREAD

**Issue by clinaresl** [zerolfx/copilot.el#60]
Hi there! 
First and foremost, thanks a lot for this! copilot.el rocks!!
I downloaded and configured copilot.el under doomemacs and it runs perfectly. I can login and also enable it in any buffer. I have tried it using different langs and I'm very satisfied with the result.
However, a key difference with copilot running in other IDEs is that it provides assistance there as the user is typing in. Here, I have to M-x copilot-complete every time I want a suggestion to pop up. 

I carefully followed the instructions given for doom emacs in the Installation paragraph, and I also verified that copilot-idle-delay is set to 0. I also updated my working copy of copilot.el and I'm currently running the very last changeset.

A typical view of *copilot events* after completing once follows (in case this helps):

[client-request] (id:6) Sun Oct 16 00:05:32 2022:
(:jsonrpc "2.0" :id 6 :method "notifyAccepted" :params
          (:uuid "d9ea9773-be62-4bf2-828c-8017b2c62de5"))
[client-request] (id:7) Sun Oct 16 00:05:32 2022:
(:jsonrpc "2.0" :id 7 :method "notifyRejected" :params
          (:uuids
           ["d9ea9773-be62-4bf2-828c-8017b2c62de5"]))
[server-reply] (id:6) Sun Oct 16 00:05:32 2022:
(:jsonrpc "2.0" :id 6 :result "OK")
[server-reply] (id:7) Sun Oct 16 00:05:32 2022:
(:jsonrpc "2.0" :id 7 :result "OK")

There is no stderr buffer so I guess this is all the information I can provide but if other is necessary I would be more than happy to provide it.

Thanks a lot!

**Comment by zerolfx** [comment]
Do you use any autosave-related plugins?

Also, please make sure `copilot-mode` is enabled.

**Comment by clinaresl** [comment]
Hey zerolfx! Thanks a lot for the prompt reply!

I forgot to mention that, yeap, copilot is enabled in the buffer with `copilot-mode` and also `M-x copilot-diagnose` returns Ok.

As for the first question, well, I have no option given in my init.el/config.el so that I assume that I have nothing out of extraordinary in doom emacs. Truth is that it seems that `auto-save-mode` is enabled by default in my buffer. However toggling it with `M-x auto-save-mode` does not help either and still copilot does not provide suggestions on the fly as I'm typing. I get'em only after `M-x copilot-complete`

I would not mind to run any tests of your interest or to provide additional information if needed ... 

Thanks a lot"!

**Comment by zerolfx** [comment]
Can you check if `copilot--post-command` is in the buffer's `post-command-hook`?
Does setting `copilot-enable-predicates` to `nil` help?

The problem mostly likely comes from `copilot--post-command` or `copilot--post-command-debounce`.

**Comment by clinaresl** [comment]
`copilot--post-command` appears indeed in the buffer's `post-command-hook`. 
On the other hand, setting `copilot-enable-predicates` to `nil` absolutely helps. Everything works perfectly all of a sudden.

Is setq'ing `copilot-enable-predicates` to `nil` in my `config.el` the only way to circumvent this? The only reason I'm asking this is to better understand what I lose by dropping the usage of `copilot-enable-predicates`

zeroflx, thanks a lot really!! Truly appreciated!

**Comment by zerolfx** [comment]
`copilot-enable-predicates` prevent Copilot from auto-completing in the following two cases:
+ Evil is enabled, and the current buffer is in normal mode.
+ The buffer is not modified. It's mainly for vanilla Emacs users to prevent Copilot from generating suggestions when navigating the code.

I guess the culprit is the second one.

**Comment by clinaresl** [comment]
Oh! I see, thanks a lot again!
That entirely closes the thread I guess

**Comment by ryankhart** [comment]
@clinaresl Would you mind pushing any latest commits that you may have for your Doom Emacs config to your repo? I can't even get this package to work in a fresh Doom Emacs config without any changes using the config code in the Doom Emacs section of this repo's README.

https://github.com/clinaresl/doom-emacs-conf

I'm about 20 hours in trying to get this to work, but I see you're on Doom Emacs and have apparently got it to work, but I can't seem to make sense of what I'm reading in this thread.

**Comment by clinaresl** [comment]
@ryankhart Sure thing Ryan! 

I do not use that repo anymore as I have now a dotfiles (private) that contains the entire configuration of my arch box. However, I just pushed the latest version of my doom config to:

 https://github.com/clinaresl/doom-emacs-conf

When using my conf files please be aware that they might require some extra installations from melpa (thought it might be easier just commenting those lines from init.el) 
Hope this helps! copilot.el works pretty nicely on doom emacs
Cheers,

**Comment by ryankhart** [comment]
@clinaresl Thanks so much! At first glance, I seem to have done what you have done already with the Copilot config. It's helpful to know that you're not having issues with it right now because this may help me pinpoint how to reproduce the bug I've been experiencing so I can report it here on a new issue thread.

It currently works great, so long as my typing cursor is not on the inside of any surrounding symbols like (), [], {}, "", '', and ``. If it is inside one of those, about half the time, hitting <tab> does an indent instead of `copilot-accept-completion` but not randomly. It seems to change and resolve itself half the time when I'm making seemingly unrelated changes to my `init.el`, `config.el`, or `package.el` files, not sure which.

I'm on MacOS, by the way.

**Comment by clinaresl** [comment]
Hi there Ryan, same behaviour observed in my archlinux. Close the buffer, open it again and you'll see the cursor locates in the same point it was before. Also, github copilot will provide the same suggestion then <TAB> works ... after a while it will fail again so closing the buffer again ... 

This is a bug, that I think the developers are already aware of ... If not, we should open a ticket

**Comment by ryankhart** [comment]
Oh that's a relief! I thought it was just me. For things like this that have a lot of config options and moving parts, I tend to assume that I'm not doing something right even if it could be considered a lack of user fault tolerance.

I am finding another workaround that may be a little more convenient than closing and reopening the buffer. And that is pressing `<esc>` to exit into normal mode, `l` to move right by 1 character, and then `<tab>` to `copilot-accept-completion`

My theory is that this may not be getting as much attention because this might not affect other language besides Elisp as much because in Elisp, it seems that literally everthing is wrapped in parentheses. And perhaps a lot of people who are trying out this package, may already have their Emacs configs settled in the way they like it.

I'll have another search through the open issues on this repo using some of the potential new terms used in this thread, and if I don't see anything, I'll open a new issue and reference this issue.

Thanks for helping with this!

**Comment by ryankhart** [comment]
Ah found it! This seems to be the exact issue I've been having.

[electric-pair and smartparens break copilot-completion-map](https://github.com/zerolfx/copilot.el/issues/86)

Once I disabled `smartparens-mode` it immediately allowed Copilot to work again.

I'll continue the discussion over in that issue's thread from now on.

**Comment by clinaresl** [comment]
Cool!!!
It was me at the beginning helping you and it finally happened that you helped me ! :)
Thanks a lot Ryan!
